chromium_src/chrome/browser/ui/views/frame/browser_frame_view_layout_linux.cc

```diff
@@ -3,14 +3,41 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#include "base/check_is_test.h"
 #include "brave/browser/ui/views/tabs/features.h"
+#include "chrome/browser/ui/views/frame/browser_view.h"
+#include "ui/views/window/caption_button_layout_constants.h"
 
 #include "src/chrome/browser/ui/views/frame/browser_frame_view_layout_linux.cc"
 
 int BrowserFrameViewLayoutLinux::NonClientTopHeight(bool restored) const {
-  if (tabs::features::ShouldShowVerticalTabs()) {
-    // Set minimum height to show caption buttons.
-    return 50;
+  if (!view_) {
+    CHECK_IS_TEST();
+    return OpaqueBrowserFrameViewLayout::NonClientTopHeight(restored);
+  }
+
+  if (view_->browser_view()->browser()->is_type_normal() &&
+      tabs::features::ShouldShowVerticalTabs()) {
+    if (!view_->ShouldShowCaptionButtons()) {
+      // In this case, the window manager might be forcibly providing system
+      // window title or it's in fullscreen mode. We shouldn't show title bar
+      // in this case.
+      return OpaqueBrowserFrameViewLayout::NonClientTopHeight(restored);
+    }
+
+    if (tabs::features::ShouldShowWindowTitleForVerticalTabs(
+            view_->browser_view()->browser())) {
+      return 30;
+    }
+
+    // TODO(sko) For now, I couldn't find a way to overlay caption buttons
```

Comment:
Were you using the GTK theme, or the classic theme? (in chrome://settings/appearance). I think with the GTK theme, the buttons should always be drawn in the title bar (no matter what). With the classic theme, I would imagine it's possible.

Reply:
I was using classing theme. Unfortunately, I couldn't put caption buttons atop toolbar :( I'm still digging on it but it's unlikely to be done soon.

Thank you for letting me know about GTK theme. It seems when selecting GTK theme, a diffrent set of frame implementation seem to be used.

For classic,
BrowserFrameViewLinux & BrowserFrameViewLayoutLinux

For GTK.
BrowserFrameViewLinuxNative & BrowserFrameViewLayoutLinuxNative

I should revisit LinuxNative set.

Reply:
Opened a folllow-up issue for this https://github.com/brave/brave-browser/issues/26233